server/src/sas/Lexer.ts

```diff
@@ -233,8 +246,277 @@ export class Lexer {
+        let multiLineStrState: false | '"""' | "'''" = false;
+        for (
+          let line = this.curr.line;
+          line < this.model.getLineCount();
+          line++
+        ) {
+          let lineContent = this.model.getLine(line);
+          if (line === this.curr.line) {
+            lineContent =
+              " ".repeat(this.curr.column) +
+              lineContent.slice(this.curr.column);
+          }
+          let pos = 0;
+          let match;
+          do {
+            if (match) {
+              pos += match.index + match[0].length;
+            }
+            if (multiLineStrState) {
+              match = /'''|"""/.exec(lineContent.substring(pos));
+              if (match && match[0] === multiLineStrState) {
+                multiLineStrState = false;
+              }
+            } else {
+              const stringReg = /'''|"""|("[^"]*?("|$))|('[^']*?('|$))/;
+              const commentReg = /#.*$/;
+              const secReg =
+                /(\b((endsubmit|endinteractive)(\s+|\/\*.*?\*\/)*;|(data|proc|%macro)\b[^'";]*;)(\s+|\/\*.*?\*\/)*$)/;
+              match = new RegExp(
+                `${stringReg.source}|${commentReg.source}|${secReg.source}`,
+                "m",
+              ).exec(lineContent.substring(pos));
+              if (match) {
+                const matchedText = match[0];
+                if (matchedText === "'''" || matchedText === '"""') {
+                  multiLineStrState = matchedText;
+                } else if (
+                  matchedText.startsWith("'") ||
+                  matchedText.startsWith('"') ||
+                  matchedText.startsWith("#")
+                ) {
+                  // do nothing to skip string and single line comment
+                } else {
+                  token = this._foundEmbeddedCodeToken(this.curr, {
+                    line: line,
+                    column: match.index,
+                  });
+                  break SWITCH;
+                }
+              }
+            }
+          } while (match);
+        }
+        token = this._foundEmbeddedCodeToken(this.curr);
+        break SWITCH;
+      }
+      case EmbeddedLangState.PROC_LUA_SUBMIT_OR_INTERACTIVE: {
+        token = this._readToken();
+        if (!token) {
+          break SWITCH;
+        }
+        if (token.type === "sep" && token.text === ";") {
+          this.context.embeddedLangState = EmbeddedLangState.PROC_LUA_CODE;
+        }
+        break SWITCH;
+      }
+      case EmbeddedLangState.PROC_LUA_CODE: {
+        let multiLineStrState: false | "[[" | "--[[" | "/*" = false;
+        for (
+          let line = this.curr.line;
+          line < this.model.getLineCount();
+          line++
+        ) {
+          let lineContent = this.model.getLine(line);
+          if (line === this.curr.line) {
+            lineContent =
+              " ".repeat(this.curr.column) +
+              lineContent.slice(this.curr.column);
+          }
```

Comment:
nit: looks like this piece of code appears multiple times. Any chance to extract it?